### ISSUE TITLE
Unit of Measure are not supported

### ISSUE THREAD

**Issue by MangelMaxime** [thoth-org/Thoth.Json#74]
I don't want to open another issue since I have another issue opened on Thoth.Fetch that I haven't fully followed up on yet, but would using it for a type that's nested about 4 layers deep count as abuse?

For me, an async computation just hangs on `Decode.Auto.fromString` without an exception or anything. Trying to decide if I should just suck it up and write a manual coder.

**Update:** It didn't like that I was using a unit of measure `<cad>` on `decimal`. I added an extra coder, and voila.

I guess UoMs aren't erased in Fable. Still not quite sure why it hangs the async though.

Update 2: Wait, my mistake, removing the UoM was the only thing that fixed it. An extra coder didn't.

_Originally posted by @reinux in https://github.com/thoth-org/Thoth.Json/issues/67#issuecomment-641596923_

**Comment by MangelMaxime** [comment]
Hello @reinux ,

I did some quick test and I think Fable has some bug when using `typeof` and UoM.

This code output `Microsoft.FSharp.Core.decimal`1`

```fs
let x = typeof<decimal<cm>>
JS.console.log(x.FullName)
```

while when using registering the extra coder via `Extra.withCustom` it seems that the generated full name is `Microsoft.FSharp.Core.decimal`1[App.View.cm]`.

So Thoth.Json output this error:

```
Error: Cannot generate auto encoder for Microsoft.FSharp.Core.decimal`1. Please pass an extra encoder.
```

It is possible to work around the problem by using a custom extra helpers and manually writing the type FullName.

```fs
open Fable.Core
open Fable.Core.JsInterop
open Thoth.Json

    // Module copied from Thoth.Json
    // I am planning to expose it in the future to avoid the need to copy/paste it in the user code base
    module internal Helpers =
        [<Emit("typeof $0")>]
        let jsTypeof (_ : JsonValue) : string = jsNative

        [<Emit("$0 instanceof SyntaxError")>]
        let isSyntaxError (_ : JsonValue) : bool = jsNative

        let inline getField (fieldName: string) (o: JsonValue) = o?(fieldName)
        let inline isString (o: JsonValue) : bool = o :? string

        let inline isBoolean (o: JsonValue) : bool = o :? bool

        let inline isNumber (o: JsonValue) : bool = jsTypeof o = "number"

        let inline isArray (o: JsonValue) : bool = JS.Constructors.Array.isArray(o)

        [<Emit("$0 === null ? false : (Object.getPrototypeOf($0 || false) === Object.prototype)")>]
        let isObject (_ : JsonValue) : bool = jsNative

        let inline isNaN (o: JsonValue) : bool = JS.Constructors.Number.isNaN(!!o)

        let inline isNullValue (o: JsonValue): bool = isNull o

        /// is the value an integer? This returns false for 1.1, NaN, Infinite, ...
        [<Emit("isFinite($0) && Math.floor($0) === $0")>]
        let isIntegralValue (_: JsonValue) : bool = jsNative

        [<Emit("$1 <= $0 && $0 < $2")>]
        let isBetweenInclusive(_v: JsonValue, _min: obj, _max: obj) = jsNative

        [<Emit("isFinite($0) && !($0 % 1)")>]
        let isIntFinite (_: JsonValue) : bool = jsNative

        let isUndefined (o: JsonValue): bool = jsTypeof o = "undefined"

        [<Emit("JSON.stringify($0, null, 4) + ''")>]
        let anyToString (_: JsonValue) : string = jsNative

        let inline isFunction (o: JsonValue) : bool = jsTypeof o = "function"

        let inline objectKeys (o: JsonValue) : string seq = upcast JS.Constructors.Object.keys(o)
        let inline asBool (o: JsonValue): bool = unbox o
        let inline asInt (o: JsonValue): int = unbox o
        let inline asFloat (o: JsonValue): float = unbox o
        let inline asFloat32 (o: JsonValue): float32 = unbox o
        let inline asString (o: JsonValue): string = unbox o
        let inline asArray (o: JsonValue): JsonValue[] = unbox o


[<Measure>] type cm

let decimalCmDecoder : Decoder<decimal<cm>> =
    fun path value ->
        if Helpers.isNumber value then
            Helpers.asFloat value |> decimal |> LanguagePrimitives.DecimalWithMeasure |> Ok
        elif Helpers.isString value then
            match System.Decimal.TryParse (Helpers.asString value) with
            | true, x -> Ok (x |> LanguagePrimitives.DecimalWithMeasure)
            | _ -> (path, BadPrimitive("a decimal", value)) |> Error
        else
            (path, BadPrimitive("a decimal", value)) |> Error

let decimalCmEncoder (v : decimal<cm>) =
    v |> string |> Encode.string

type Data =
    {
        Name : string
        Value : decimal<cm>
    }

let value =
    {
        Name = "Centimeters"
        Value = 3.0m<cm>
    }

// Copied from Thoth.Json in order to debug the retrieve type
let inline withCustomAndKeyAndTypeFullName (typeFullName : string) (encoder: Encoder<'Value>) (decoder: Decoder<'Value>)
           (extra: ExtraCoders): ExtraCoders =
    JS.console.log(typeof<'Value>.FullName)
    { extra with
          Hash = System.Guid.NewGuid().ToString()
          Coders =
              extra.Coders |> Map.add typeFullName (Encode.boxEncoder encoder, Decode.boxDecoder decoder) }

let extra =
    Extra.empty
    |> withCustomAndKeyAndTypeFullName "Microsoft.FSharp.Core.decimal`1" decimalCmEncoder decimalCmDecoder

let json =
    Encode.Auto.toString(4, value, extra = extra)

match Decode.Auto.fromString(json, extra = extra) with
| Ok decoded ->
    JS.console.log decoded.Name
    JS.console.log decoded.Value
| Error err ->
    JS.console.error err
```

Just to be clear, there is definitly a bug in Fable (or Thoth.Json). I am providing you with this code so you can be unblocked.

@alfonsogarciacaro I am pinging you just in case this problem ring something to you. I will try to provide a minimal reproduction code outside of Thoth.Json.

**Comment by reinux** [comment]
Thanks, I appreciate it!

I didn't think there could be a workaround for this.

**Comment by alfonsogarciacaro** [comment]
@MangelMaxime Yes, there's an issue now with units of measure reflection right now in Fable. I need to fix it, sorry!

**Comment by sWW26** [comment]
Hi @alfonsogarciacaro we're experiencing this with Fable 3.  Is this still an known issue?  I had a quick look in the Fable issues and couldn't see anything open that's related.

**Comment by reinux** [comment]
I ran into it recently too, so I'm assuming it's on the backburner.

I'm using a simpler around for now, just wrapping the type in a single-case union:

```
type CAD =
  | CAD of decimal
  member x.ToUoM = let (CAD d) = x in d * 1m<cad>
  static member FromUoM(x: decimal<cad>) = CAD(x / 1m<cad>)
```

**Comment by alfonsogarciacaro** [comment]
Ups, we fixed several issues with units of measure but seems in64/decimal fell off the radar here. Can you please try with dotnet fable 3.1.1?

**Comment by reinux** [comment]
Hey hey, it works! Thanks, @alfonsogarciacaro

**Comment by MangelMaxime** [comment]
Closing as Fable 3.1.1 is fixing this problem.
